File: EuclideanClusterExtraction.lua
local ffi = require 'ffi'
local torch = require 'torch'
local utils = require 'pcl.utils'
local pcl = require 'pcl.PointTypes'

local EuclideanClusterExtraction = torch.class('pcl.EuclideanClusterExtraction', pcl)
local EuclideanClusterExtraction_func_by_type = {}

local function init()
  local EuclideanClusterExtraction_method_names = {
    'new',
    'delete',
    'EuclideanClusterExtraction_ptr',
    'setInputCloud',
    'setIndices',
    'setSearchMethod_Octree',
    'setSearchMethod_KdTree',
    'setClusterTolerance',
    'getClusterTolerance',
    'setMinClusterSize',
    'getMinClusterSize',
    'setMaxClusterSize',
    'getMaxClusterSize',
    'extract'
  }
  for k,v in pairs(utils.type_key_map) do
    EuclideanClusterExtraction_func_by_type[k] = utils.create_typed_methods("pcl_EuclideanClusterExtraction_TYPE_KEY_", EuclideanClusterExtraction_method_names, v)
  end
end

init()

function EuclideanClusterExtraction:__init(pointType)
  self.pointType = pcl.pointType(pointType or pcl.PointXYZ)
  self.f = EuclideanClusterExtraction_func_by_type[self.pointType]
  self.h = self.f.new()
  self.p = self.f.EuclideanClusterExtraction_ptr(self.h)
end

function EuclideanClusterExtraction:handle()
  return self.h
end

function EuclideanClusterExtraction:EuclideanClusterExtraction_ptr()
  return self.p
end

function EuclideanClusterExtraction:setInputCloud(cloud)
  self.f.setInputCloud(self.p, cloud:cdata())
end

function EuclideanClusterExtraction:setIndices(indices)
  self.f.setIndices(self.p, indices:cdata())
end

function EuclideanClusterExtraction:setSearchMethod(search)
  if torch.isTypeOf(search, pcl.KdTree) then
    self.f.setSearchMethod_KdTree(self.p, search:cdata())
  elseif torch.isTypeOf(search, pcl.Octree) then
    self.f.setSearchMethod_Octree(self.p, search:cdata())
  else
    error("unsupported search method")
  end
end

function EuclideanClusterExtraction:setClusterTolerance(cluster_tolerance)
  self.f.setClusterTolerance(self.p, cluster_tolerance)
end

function EuclideanClusterExtraction:getClusterTolerance()
  return self.f.getClusterTolerance(self.p)
end

function EuclideanClusterExtraction:setMinClusterSize(min_cluster_size)
  self.f.setMinClusterSize(self.p, min_cluster_size)
end

function EuclideanClusterExtraction:getMinClusterSize()
  return self.f.getMinClusterSize(self.p)
end

function EuclideanClusterExtraction:setMaxClusterSize(max_cluster_size)
  self.f.setMaxClusterSize(self.p, max_cluster_size)
end

function EuclideanClusterExtraction:getMaxClusterSize()
  return self.f.getMaxClusterSize(self.p)
end

function EuclideanClusterExtraction:extract(clusters)
  clusters = clusters or pcl.IndicesVector()
  self.f.extract(self.p, clusters:cdata())
  return clusters
end

File: src/generic/euclidean_cluster_extraction.cpp
#include "searchwrapper.h"
#include <pcl/segmentation/extract_clusters.h>

#define EuclideanClusterExtraction_ pcl::EuclideanClusterExtraction<_PointT>
#define EuclideanClusterExtraction_handle boost::shared_ptr<EuclideanClusterExtraction_>

#define PointCloud_ptr pcl::PointCloud<_PointT>::Ptr
#define KdTree_ptr pcl::KdTreeFLANN<_PointT>::Ptr
#define Octree_ptr pcl::octree::OctreePointCloudSearch<_PointT>::Ptr

PCLIMP(EuclideanClusterExtraction_handle *, EuclideanClusterExtraction, new)() {
  return new EuclideanClusterExtraction_handle(new EuclideanClusterExtraction_());
}

PCLIMP(void, EuclideanClusterExtraction, delete)(EuclideanClusterExtraction_handle *handle) {
  delete handle;
}

PCLIMP(EuclideanClusterExtraction_ *, EuclideanClusterExtraction, EuclideanClusterExtraction_ptr)(EuclideanClusterExtraction_handle *handle) {
  return handle->get();
}

// PCLBase methods
PCLIMP(void, EuclideanClusterExtraction, setInputCloud)(EuclideanClusterExtraction_ *self, PointCloud_ptr *cloud) {
  self->setInputCloud(*cloud);
}

PCLIMP(void, EuclideanClusterExtraction, setIndices)(EuclideanClusterExtraction_ *self, Indices_ptr *indices) {
  self->setIndices(*indices);
}

PCLIMP(void, EuclideanClusterExtraction, setSearchMethod_Octree)(EuclideanClusterExtraction_ *self, Octree_ptr *octree) {
  boost::shared_ptr<OctreeSearchWrapper<_PointT> > wrapper(new OctreeSearchWrapper<_PointT>(*octree));
  self->setSearchMethod(wrapper);
}

PCLIMP(void, EuclideanClusterExtraction, setSearchMethod_KdTree)(EuclideanClusterExtraction_ *self, KdTree_ptr *kdtree) {
  boost::shared_ptr<KdTreeSearchWrapper<_PointT> > wrapper(new KdTreeSearchWrapper<_PointT>(*kdtree));
  self->setSearchMethod(wrapper);
}

PCLIMP(void, EuclideanClusterExtraction, setClusterTolerance)(EuclideanClusterExtraction_ *self, double tolerance) {
  self->setClusterTolerance(tolerance);
}

PCLIMP(double, EuclideanClusterExtraction, getClusterTolerance)(EuclideanClusterExtraction_ *self) {
  return self->getClusterTolerance();
}

PCLIMP(void, EuclideanClusterExtraction, setMinClusterSize)(EuclideanClusterExtraction_ *self, int min_cluster_size) {
  self->setMinClusterSize(min_cluster_size);
}

PCLIMP(int, EuclideanClusterExtraction, getMinClusterSize)(EuclideanClusterExtraction_ *self) {
  return self->getMinClusterSize();
}

PCLIMP(void, EuclideanClusterExtraction, setMaxClusterSize)(EuclideanClusterExtraction_ *self, int max_cluster_size) {
  self->setMaxClusterSize(max_cluster_size);
}

PCLIMP(int, EuclideanClusterExtraction, getMaxClusterSize)(EuclideanClusterExtraction_ *self) {
  return self->getMaxClusterSize();
}

PCLIMP(void, EuclideanClusterExtraction, extract)(EuclideanClusterExtraction_ *self, IndicesVector_ptr *clusters) {
  std::vector<pcl::PointIndices> _clusters;
  self->extract(_clusters);

  // copy indices
  (*clusters)->clear();
  for (std::vector<pcl::PointIndices>::const_iterator i = _clusters.begin(); i != _clusters.end(); ++i)
    (*clusters)->push_back(Indices_ptr(new std::vector<int>(i->indices)));
}

#undef EuclideanClusterExtraction_
#undef EuclideanClusterExtraction_handle
#undef PointCloud_ptr
#undef KdTree_ptr
#undef Octree_ptr
